compiler/rustc_data_structures/src/sharded.rs

```diff
@@ -73,6 +73,44 @@ impl<T> Sharded<T> {
         }
     }
 
+    /// The shard is selected by hashing `val` with `FxHasher`.
+    #[inline]
+    pub fn lock_shard_by_value<K: Hash + ?Sized>(&self, _val: &K) -> LockGuard<'_, T> {
+        match self {
+            Self::Single(single) => {
+                // SAFETY: We know `is_dyn_thread_safe` was false when creating the lock thus
+                // `might_be_dyn_thread_safe` was also false.
```

Comment:
This comment doesn't explain the relation between `Sharded` and these two functions. I think it's easier to understand if we say `this is under single thread` directly.

Reply:
That doesn't sound sufficiently precise to explain the precondition on `lock_assume_no_sync`. `might_be_dyn_thread_safe` may not always match `is_dyn_thread_safe`.

Reply:
I added an additional higher level comment here.

Reply:
OK